internal/infrastructure/live-store/redis/round.go

```diff
@@ -182,3 +189,160 @@ func (b *boardingInputsStore) Get() int {
 	}
 	return num
 }
+
+func (b *boardingInputsStore) AddSignatures(
+	ctx context.Context, batchId string, inputSigs map[uint32]ports.SignedBoardingInput,
+) (err error) {
+	key := fmt.Sprintf("%s:%s", boardingInputSigsKey, batchId)
+
+	// Prepare arguments first so serialization errors happen before the transaction
+	type fieldVal struct {
+		field string
+		value string
+	}
+	fields := make([]fieldVal, 0, len(inputSigs))
+
+	for inIndex, sig := range inputSigs {
+		field := fmt.Sprintf("%d", inIndex)
+		value, err := newSigsDTO(sig).serialize()
+		if err != nil {
+			return err
+		}
+		fields = append(fields, fieldVal{field, string(value)})
+	}
+
+	// Transactional update with retry logic
+	for range b.numOfRetries {
+		if err = b.rdb.Watch(ctx, func(tx *redis.Tx) error {
+			_, err := tx.TxPipelined(ctx, func(pipe redis.Pipeliner) error {
+				for _, fv := range fields {
+					pipe.HSetNX(ctx, key, fv.field, fv.value)
+				}
+				return nil
+			})
+			return err
```

Comment:
```suggestion
			}, key)
			return err
```

Just one note on Redis transactions: here we use Watch(ctx, fn, key) + TxPipelined. Watch adds optimistic locking on the given key, and inside the callback TxPipelined wraps the commands in a MULTI/EXEC block so they execute as a single atomic batch if the watched key hasn’t changed; if it has changed, Redis aborts EXEC and go-redis returns redis.TxFailedErr, which we can retry. Unlike SQL transactions, Redis does not provide full rollback semantics: commands inside EXEC are not undone if one fails or if something like HSETNX just returns 0 for an existing field – the batch is atomic with respect to other clients, but it’s not all-or-nothing in the SQL sense.